<div class="vtbegenerated">
 <p><span style="font-size: medium;"><strong>Who Survived the Titanic?</strong></span></p> 
 <p>For this assignment, we will analyze the open dataset with real data on the passengers aboard the Titanic.</p> 
 <p>Download the data from Kaggle website: <a href="https://www.kaggle.com/c/titanic/data" target="_blank">https://www.kaggle.com/c/titanic/data</a>, file <strong>“train.csv”.&nbsp;</strong>The definition of all variables can be found on the same page, in the Data Dictionary section.</p> 
 <p>Read the data from the file into pandas dataframe.&nbsp;Analyze, clean and transform the data to answer the following question:</p> 
 <ul> 
  <li><strong>What categories of passengers were most likely to survive the Titanic disaster?</strong></li> 
 </ul> 
 <p>You might include the following attributes in your analysis:</p> 
 <ul> 
  <li>Passenger age,</li> 
  <li>Passenger gender,</li> 
  <li>Cabin class the passenger travelled in (variable ‘ticket class’).</li> 
 </ul> 
 <p>What other attributes did you use for the analysis? Explain how you used them.Provide a complete list of all attributes used.</p> 
 <p>Did you engineer any attributes? If yes, explain the rationale and how the new attributes were used in the analysis?</p> 
 <p>If you have excluded any attributes from the analysis, provide an explanation why you believe they can be excluded.</p> 
 <p>How did you treat missing values? Provide a&nbsp;detailed explanation in the comments.</p> 
 <br> 
 <br>
</div>

<div class="markdown-content-box__converter"><div class="markdown-converter__text--rendered competition-data__content">
<h3>Data Dictionary (taken from https://www.kaggle.com/c/titanic/data):</h3>
<table style="width: 100%;">
<tbody>
<tr><th><b>Variable</b></th><th><b>Definition</b></th><th><b>Key</b></th></tr>
<tr>
<td>survival</td>
<td>Survival</td>
<td>0 = No, 1 = Yes</td>
</tr>
<tr>
<td>pclass</td>
<td>Ticket class</td>
<td>1 = 1st, 2 = 2nd, 3 = 3rd</td>
</tr>
<tr>
<td>sex</td>
<td>Sex</td>
<td></td>
</tr>
<tr>
<td>Age</td>
<td>Age in years</td>
<td></td>
</tr>
<tr>
<td>sibsp</td>
<td># of siblings / spouses aboard the Titanic</td>
<td></td>
</tr>
<tr>
<td>parch</td>
<td># of parents / children aboard the Titanic</td>
<td></td>
</tr>
<tr>
<td>ticket</td>
<td>Ticket number</td>
<td></td>
</tr>
<tr>
<td>fare</td>
<td>Passenger fare</td>
<td></td>
</tr>
<tr>
<td>cabin</td>
<td>Cabin number</td>
<td></td>
</tr>
<tr>
<td>embarked</td>
<td>Port of Embarkation</td>
<td>C = Cherbourg, Q = Queenstown, S = Southampton</td>
</tr>
</tbody>
</table>
<h3>Variable Notes</h3>
<p><b>pclass</b>: A proxy for socio-economic status (SES)<br> 1st = Upper<br> 2nd = Middle<br> 3rd = Lower<br><br> <b>age</b>: Age is fractional if less than 1. If the age is estimated, is it in the form of xx.5<br><br> <b>sibsp</b>: The dataset defines family relations in this way...<br> Sibling = brother, sister, stepbrother, stepsister<br> Spouse = husband, wife (mistresses and fiancés were ignored)<br><br> <b>parch</b>: The dataset defines family relations in this way...<br> Parent = mother, father<br> Child = daughter, son, stepdaughter, stepson<br> Some children travelled only with a nanny, therefore parch=0 for them.</p></div></div>

In [1]:
import pandas as pd
import numpy as np
import csv

Read the Titanic passenger data from a .csv file into a DataFrame for analysis and output the contents for review.

In [2]:
df = pd.read_csv('train.csv', index_col=0)
df

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
6                   0       3   
7                   0       1   
8                   0       3   
9                   1       3   
10                  1       2   
11                  1       3   
12                  1       1   
13                  0       3   
14                  0       3   
15                  0       3   
16                  1       2   
17                  0       3   
18                  1       2   
19                  0       3   
20                  1       3   
21                  0       2   
22                  1       2   
23                  1       3   
24                  1       1   
25                  0       3   
26                  1       3   
27                  0       3   
28                  0       1   
29                  1       3   
30                  0       3   
...               ...     ...   
862                 0       2   
863                 1       1   
864                 0       3   
865                 0       2   
866                 1       2   
867                 1       2   
868                 0       1   
869                 0       3   
870                 1       3   
871                 0       3   
872                 1       1   
873                 0       1   
874                 0       3   
875                 1       2   
876                 1       3   
877                 0       3   
878                 0       3   
879                 0       3   
880                 1       1   
881                 1       2   
882                 0       3   
883                 0       3   
884                 0       2   
885                 0       3   
886                 0       3   
887                 0       2   
888                 1       1   
889                 0       3   
890                 1       1   
891                 0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
6                                             Moran, Mr. James    male   NaN   
7                                      McCarthy, Mr. Timothy J    male  54.0   
8                               Palsson, Master. Gosta Leonard    male   2.0   
9            Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0   
10                         Nasser, Mrs. Nicholas (Adele Achem)  female  14.0   
11                             Sandstrom, Miss. Marguerite Rut  female   4.0   
12                                    Bonnell, Miss. Elizabeth  female  58.0   
13                              Saundercock, Mr. William Henry    male  20.0   
14                                 Andersson, Mr. Anders Johan    male  39.0   
15                        Vestrom, Miss. Hulda Amanda Adolfina  female  14.0   
16                            Hewlett, Mrs. (Mary D Kingcome)   female  55.0   
17                                        Rice, Master. Eugene    male   2.0   
18                                Williams, Mr. Charles Eugene    male   NaN   
19           Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0   
20                                     Masselmani, Mrs. Fatima  female   NaN   
21                                        Fynney, Mr. Joseph J    male  35.0   
22                                       Beesley, Mr. Lawrence    male  34.0   
23           

Let's find out what kind of data we're dealing with:

In [3]:
df.dtypes

Survived      int64
Pclass        int64
Name         object
Sex          object
Age         float64
SibSp         int64
Parch         int64
Ticket       object
Fare        float64
Cabin        object
Embarked     object
dtype: object

Let's see what the numbers in the data set tell us.

In [4]:
df.describe()

Survived      Pclass         Age       SibSp       Parch        Fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

In [5]:
((891-714)/891)*100

19.865319865319865

We can see from the above that we are missing about 20% of our Age values, because it only counted 714 values in the 891 rows. That means 177 values in the Age column are NaN. We can also see that the majority of the passengers whose age is known, are under 38 or younger, the youngest passenger is less than 1 year old and the oldest passenger is 80 years old.

The describe() function only provided information about the numeric typed columns in the DataFrame. What about the non-numeric columns?

In [6]:
df.describe(exclude=[np.number])

Name   Sex  Ticket Cabin Embarked
count                            891   891     891   204      889
unique                           891     2     681   147        3
top     Skoog, Master. Karl Thorsten  male  347082    G6        S
freq                               1   577       7     4      644

This shows us that these attributes are very well populated, with the exception of Cabin. There are many missing values in this attribute. Cabin COULD be useful as a way to group passengers to determine if they are travelling together. But in this data set, Ticket could serve the same purpose much more effectively, since it is fully populated. Thus, Cabin isn't particularly useful to us and can be dropped from the DataFrame. As we will see further below, the Pclass attribute will be more useful to us as a socio-economic grouping than anything we would derive from the Fare attribute, so we will drop that as well.

In [7]:
df2 = df.drop(['Cabin','Fare'], axis=1)
df2

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
6                   0       3   
7                   0       1   
8                   0       3   
9                   1       3   
10                  1       2   
11                  1       3   
12                  1       1   
13                  0       3   
14                  0       3   
15                  0       3   
16                  1       2   
17                  0       3   
18                  1       2   
19                  0       3   
20                  1       3   
21                  0       2   
22                  1       2   
23                  1       3   
24                  1       1   
25                  0       3   
26                  1       3   
27                  0       3   
28                  0       1   
29                  1       3   
30                  0       3   
...               ...     ...   
862                 0       2   
863                 1       1   
864                 0       3   
865                 0       2   
866                 1       2   
867                 1       2   
868                 0       1   
869                 0       3   
870                 1       3   
871                 0       3   
872                 1       1   
873                 0       1   
874                 0       3   
875                 1       2   
876                 1       3   
877                 0       3   
878                 0       3   
879                 0       3   
880                 1       1   
881                 1       2   
882                 0       3   
883                 0       3   
884                 0       2   
885                 0       3   
886                 0       3   
887                 0       2   
888                 1       1   
889                 0       3   
890                 1       1   
891                 0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
6                                             Moran, Mr. James    male   NaN   
7                                      McCarthy, Mr. Timothy J    male  54.0   
8                               Palsson, Master. Gosta Leonard    male   2.0   
9            Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0   
10                         Nasser, Mrs. Nicholas (Adele Achem)  female  14.0   
11                             Sandstrom, Miss. Marguerite Rut  female   4.0   
12                                    Bonnell, Miss. Elizabeth  female  58.0   
13                              Saundercock, Mr. William Henry    male  20.0   
14                                 Andersson, Mr. Anders Johan    male  39.0   
15                        Vestrom, Miss. Hulda Amanda Adolfina  female  14.0   
16                            Hewlett, Mrs. (Mary D Kingcome)   female  55.0   
17                                        Rice, Master. Eugene    male   2.0   
18                                Williams, Mr. Charles Eugene    male   NaN   
19           Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0   
20                                     Masselmani, Mrs. Fatima  female   NaN   
21                                        Fynney, Mr. Joseph J    male  35.0   
22                                       Beesley, Mr. Lawrence    male  34.0   
23           

The data is starting to look pretty good at this point. We've indexed the data on PassengerID and dropped the Cabin and Fare columns. Age is missing some values, but it has enough data in it to still serve a purpose. Let's start to look at how we can categorize the passengers. One of the more obvious categories is gender. What is the male/female split among the passengers?

In [8]:
df2.groupby(by='Sex').Sex.count()

Sex
female    314
male      577
Name: Sex, dtype: int64

We now know that there was almost twice as many males on-board as there was females. From this we might assume that more males would have survived. But is that what the data says?

Count the number of surviving and deceased passengers by gender

In [9]:
df2.groupby(by=['Survived','Sex']).Sex.count()

Survived  Sex   
0         female     81
          male      468
1         female    233
          male      109
Name: Sex, dtype: int64

This showed that despite the fact that there was twice as many males on board, almost 6x as many males died as females. And there were twice as many female survivors as male survivors. So it's pretty clear which gender was most likely to survive. But we have other data too...

Pclass is provided as an indicator of socio-economic status. It contains 3 distinct integer values, 1, 2 and 3 to indicate Upper, Middle and Lower class respectively. It may be useful to indicate this status using a more descriptive value.

In [10]:
classType = {1:'Upper', 2:'Middle', 3:'Lower'}

def getSES(typecd):
    return classType[typecd]

df2['SES'] = df2.apply(lambda row: getSES(row['Pclass']), axis=1)
df2

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
6                   0       3   
7                   0       1   
8                   0       3   
9                   1       3   
10                  1       2   
11                  1       3   
12                  1       1   
13                  0       3   
14                  0       3   
15                  0       3   
16                  1       2   
17                  0       3   
18                  1       2   
19                  0       3   
20                  1       3   
21                  0       2   
22                  1       2   
23                  1       3   
24                  1       1   
25                  0       3   
26                  1       3   
27                  0       3   
28                  0       1   
29                  1       3   
30                  0       3   
...               ...     ...   
862                 0       2   
863                 1       1   
864                 0       3   
865                 0       2   
866                 1       2   
867                 1       2   
868                 0       1   
869                 0       3   
870                 1       3   
871                 0       3   
872                 1       1   
873                 0       1   
874                 0       3   
875                 1       2   
876                 1       3   
877                 0       3   
878                 0       3   
879                 0       3   
880                 1       1   
881                 1       2   
882                 0       3   
883                 0       3   
884                 0       2   
885                 0       3   
886                 0       3   
887                 0       2   
888                 1       1   
889                 0       3   
890                 1       1   
891                 0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
6                                             Moran, Mr. James    male   NaN   
7                                      McCarthy, Mr. Timothy J    male  54.0   
8                               Palsson, Master. Gosta Leonard    male   2.0   
9            Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0   
10                         Nasser, Mrs. Nicholas (Adele Achem)  female  14.0   
11                             Sandstrom, Miss. Marguerite Rut  female   4.0   
12                                    Bonnell, Miss. Elizabeth  female  58.0   
13                              Saundercock, Mr. William Henry    male  20.0   
14                                 Andersson, Mr. Anders Johan    male  39.0   
15                        Vestrom, Miss. Hulda Amanda Adolfina  female  14.0   
16                            Hewlett, Mrs. (Mary D Kingcome)   female  55.0   
17                                        Rice, Master. Eugene    male   2.0   
18                                Williams, Mr. Charles Eugene    male   NaN   
19           Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0   
20                                     Masselmani, Mrs. Fatima  female   NaN   
21                                        Fynney, Mr. Joseph J    male  35.0   
22                                       Beesley, Mr. Lawrence    male  34.0   
23           

We can look for survivors based on SES.

In [11]:
df2.groupby(by=['Survived','SES']).SES.count()

Survived  SES   
0         Lower     372
          Middle     97
          Upper      80
1         Lower     119
          Middle     87
          Upper     136
Name: SES, dtype: int64

This clearly shows that a disproportionate number of lower class passengers died relative to the other socio-economic groups. We could have run the same grouped count using Pclass, but SES gave us more descriptive output.

Now we can combine dimensions of our analysis for additional insight:

In [12]:
df2.groupby(by=['Survived','Sex','SES']).SES.count()

Survived  Sex     SES   
0         female  Lower      72
                  Middle      6
                  Upper       3
          male    Lower     300
                  Middle     91
                  Upper      77
1         female  Lower      72
                  Middle     70
                  Upper      91
          male    Lower      47
                  Middle     17
                  Upper      45
Name: SES, dtype: int64

Looking at upper class females, 91 survived and 3 died.<br/>
Looking at middle class females,70 survived and 6 died.<br/>
Looking at lower class females, 72 survived and <b><u>72</u> died!</b> Socio-economic status clearly impacted passenger survival on the Titanic.

The Embarked attribute represents the port of embarkation. It contains only 3 valid values, C, Q and S:<br/>
C = Cherbourg, Q = Queenstown, S = Southampton

Similar to what we did with the Pclass attribute, we can assign more descriptive values that represent the same data, which will improve the quality of our output. We know already from above that the Embarked attribute isn't full populated, but it is only missing 2 values in the whole set, which should not impact any results significantly. Our grouped count calculation ignores null values anyway:

In [13]:
port = {'C':'Cherbourg', 'Q':'Queenstown', 'S':'Southampton'}

def getPort(portcd):
    if not pd.isnull(portcd):
        return port[portcd]

df2['Port'] = df2.apply(lambda row: getPort(row['Embarked']), axis=1)
df2

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
6                   0       3   
7                   0       1   
8                   0       3   
9                   1       3   
10                  1       2   
11                  1       3   
12                  1       1   
13                  0       3   
14                  0       3   
15                  0       3   
16                  1       2   
17                  0       3   
18                  1       2   
19                  0       3   
20                  1       3   
21                  0       2   
22                  1       2   
23                  1       3   
24                  1       1   
25                  0       3   
26                  1       3   
27                  0       3   
28                  0       1   
29                  1       3   
30                  0       3   
...               ...     ...   
862                 0       2   
863                 1       1   
864                 0       3   
865                 0       2   
866                 1       2   
867                 1       2   
868                 0       1   
869                 0       3   
870                 1       3   
871                 0       3   
872                 1       1   
873                 0       1   
874                 0       3   
875                 1       2   
876                 1       3   
877                 0       3   
878                 0       3   
879                 0       3   
880                 1       1   
881                 1       2   
882                 0       3   
883                 0       3   
884                 0       2   
885                 0       3   
886                 0       3   
887                 0       2   
888                 1       1   
889                 0       3   
890                 1       1   
891                 0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
6                                             Moran, Mr. James    male   NaN   
7                                      McCarthy, Mr. Timothy J    male  54.0   
8                               Palsson, Master. Gosta Leonard    male   2.0   
9            Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0   
10                         Nasser, Mrs. Nicholas (Adele Achem)  female  14.0   
11                             Sandstrom, Miss. Marguerite Rut  female   4.0   
12                                    Bonnell, Miss. Elizabeth  female  58.0   
13                              Saundercock, Mr. William Henry    male  20.0   
14                                 Andersson, Mr. Anders Johan    male  39.0   
15                        Vestrom, Miss. Hulda Amanda Adolfina  female  14.0   
16                            Hewlett, Mrs. (Mary D Kingcome)   female  55.0   
17                                        Rice, Master. Eugene    male   2.0   
18                                Williams, Mr. Charles Eugene    male   NaN   
19           Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0   
20                                     Masselmani, Mrs. Fatima  female   NaN   
21                                        Fynney, Mr. Joseph J    male  35.0   
22                                       Beesley, Mr. Lawrence    male  34.0   
23           

Now we can look for a relationship between port of embarkment and survival:

In [14]:
df2.groupby(by=['Survived','Port']).Port.count()

Survived  Port       
0         Cherbourg       75
          Queenstown      47
          Southampton    427
1         Cherbourg       93
          Queenstown      30
          Southampton    217
Name: Port, dtype: int64

In [15]:
[(93/(93+75)*100),(30/(30+47)*100),(217/(217+427))*100]

[55.35714285714286, 38.961038961038966, 33.69565217391305]

Cherbourg survival: 55%, Queenstown survival: 38%, Southampton survival: 33%

A higher percentage of passengers from Southampton survived than Queenstown or Cherbourg. Based on what we've already learned, this may tell us that Southampton is more affluent than Queenstown and Cherbourg. Or it may tell us that more women embarked from Southampton than from the other 2 ports. We can test those assumptions:

In [16]:
womenSurvivorsByPort = df2[(df2['Survived'] == 1) & (df2['Sex']=='female')].groupby(by=['Port']).Port.count()
womenSurvivorsByPort

Port
Cherbourg       64
Queenstown      27
Southampton    140
Name: Port, dtype: int64

These results validate the assumption that more women embarked from Southampton.

Now to test if Southampton is the more affluent port:

In [17]:
df2.groupby(by=['Port','SES']).SES.count()

Port         SES   
Cherbourg    Lower      66
             Middle     17
             Upper      85
Queenstown   Lower      72
             Middle      3
             Upper       2
Southampton  Lower     353
             Middle    164
             Upper     127
Name: SES, dtype: int64

Our second assumption was correct as well. Well partially anyway. It's true that there were far more Upper and Middle Class passengers from Southampton, than from Queenstown or Cherbourg. But there were also far more Lower class passengers from Southampton. It's probably more accurate to just say that there were far more passengers embarking from Southampton than from the other 2 ports.

So far we have learned that both gender and socio-economic status greatly influence who survived the disaster. We have also learned that more passengers embarked from Southampton than the other 2 ports, so it follows that the majority of the survivors were from Southampton. Another variable we can examine is age. Were children more likely to die since they have less strength than adults or because they haven't learned to swim yet? Or were children more likely to survive because room was made for them in the life boats? Were the elderly cared for in favour of younger adults who were more capable of looking after themselves? We can look for answers in the data to determine if a passenger's age was a factor in their survival or demise. But as we learned above, we are missing about 20% of our age values. One approach could be to just discard those records and not consider those passengers in the analysis. But the more data
we have, the more proof we have to support our conclusions.

Even if an age value is missing, there may be clues in our passenger records which give us an indication of their relative age. Some of the clues I have found include:
1. A female who uses the "Mrs." honorific is either married, divorced or a widow. In those cases, she is likely an adult.
2. A male who uses the "Master." honorific is a child. This honorific is typically only used for male children.
3. A male who uses the "Mr." honorific is an adult. If they were a child, "Master." would be used.
4. A passenger who is travelling alone is most likely not a child, thus an adult.

We will engineer an "AgeCategory" attribute, whose valid values are "Child", "Adult" or None (for unknown) and use it as a dimension in our analysis.

For starters, Age is typically the best indicator of whether or not a person is a child or an adult. But what does "adult" mean? Biological adulthood starts in the early teens or before. Legal adulthood in England in 1912 was 21 years of age. For the purpose of this analysis, we will split the difference and consider 16 to be the age of adulthood and anyone younger would be considered a child.

By default, set all AgeCategory values to None. We can update the attribute as we learn more about the passengers.

In [18]:
df2['AgeCategory'] = None

Now apply the age test to determine AgeCategory

In [19]:
df2.loc[df2['Age'] >= 16, ['AgeCategory']] = 'Adult'
df2.loc[df2['Age'] < 16, ['AgeCategory']] = 'Child'

For passengers whose age we did not know, we will use the other clues discussed to derive AgeCategory.

In [20]:
# If Mr./Mrs. honorific is used in the name, AgeCategory = 'Adult'
df2.loc[df2['Name'].str.contains('Mrs\.|Mr\.'), ['AgeCategory']] = 'Adult'

# If Master. honorific is used in the name, AgeCategory = 'Child'
df2.loc[df2['Name'].str.contains('Master\.'), ['AgeCategory']] = 'Child'

# For each passenger, calculate the number of travelling companions
df2['Companions'] = df.SibSp + df.Parch

# If Companions = 0, AgeCategory = 'Adult'
df2.loc[df2['Companions'] == 0, ['AgeCategory']] = 'Adult'

Let's take a look at how many passengers still do not have a value for AgeCategory

In [21]:
df2[df2['AgeCategory'].isnull()]

Survived  Pclass                                      Name  \
PassengerId                                                               
110                 1       3                       Moran, Miss. Bertha   
129                 1       3                         Peter, Miss. Anna   
181                 0       3              Sage, Miss. Constance Gladys   
230                 0       3                   Lefebre, Miss. Mathilde   
241                 0       3                     Zabour, Miss. Thamine   
242                 1       3            Murphy, Miss. Katherine "Kate"   
331                 1       3                        McCoy, Miss. Agnes   
410                 0       3                        Lefebre, Miss. Ida   
486                 0       3                    Lefebre, Miss. Jeannie   
594                 0       3                        Bourke, Miss. Mary   
613                 1       3               Murphy, Miss. Margaret Jane   
793                 0       3                   Sage, Miss. Stella Anna   
864                 0       3         Sage, Miss. Dorothy Edith "Dolly"   
889                 0       3  Johnston, Miss. Catherine Helen "Carrie"   

                Sex  Age  SibSp  Parch      Ticket Embarked    SES  \
PassengerId                                                          
110          female  NaN      1      0      371110        Q  Lower   
129          female  NaN      1      1        2668        C  Lower   
181          female  NaN      8      2    CA. 2343        S  Lower   
230          female  NaN      3      1        4133        S  Lower   
241          female  NaN      1      0        2665        C  Lower   
242          female  NaN      1      0      367230        Q  Lower   
331          female  NaN      2      0      367226        Q  Lower   
410          female  NaN      3      1        4133        S  Lower   
486          female  NaN      3      1        4133        S  Lower   
594          female  NaN      0      2      364848        Q  Lower   
613          female  NaN      1      0      367230        Q  Lower   
793          female  NaN      8      2    CA. 2343        S  Lower   
864          female  NaN      8      2    CA. 2343        S  Lower   
889          female  NaN      1      2  W./C. 6607        S  Lower   

                    Port AgeCategory  Companions  
PassengerId                                       
110           Queenstown        None           1  
129            Cherbourg        None           2  
181          Southampton        None          10  
230          Southampton        None           4  
241            Cherbourg        None           1  
242           Queenstown        None           1  
331           Queenstown        None           2  
410          Southampton        None           4  
486          Southampton        None           4  
594           Queenstown        None           2  
613           Queenstown        None           1  
793          Southampton        None          10  
864          Southampton        None          10  
889          Southampton        None           3

We now have an age category for all but 14 passengers, which is very good considering we are missing an age value for 177 passengers. All of these remaining 14 passengers use the honorific 'Miss.', which is not a strong indicator of relative age. It only implies a female's marital status of being single, which could be a female of any age. Since we have no data left to indicate the age category for these passengers, and it is an insignificant number of records relative to the total number, we can drop these passengers from our age analysis.

In [22]:
df2.drop(df2[df2.AgeCategory.isnull()].index, axis=0, inplace=True)

Now we can look for survivors based on their age category

In [23]:
df2.groupby(by=['Survived','AgeCategory']).AgeCategory.count()

Survived  AgeCategory
0         Adult          509
          Child           31
1         Adult          292
          Child           45
Name: AgeCategory, dtype: int64

What percentage of children survived?

In [24]:
45/(45+31)*100

59.210526315789465

What percentage of adults survived?

In [25]:
292/(292+509)*100

36.45443196004994

From the above we can see that a greater percentage of children (59%) survived than adults (36%).

The following section summarizes our findings to address the original questions posed by the rubric for the assignment:

<b>QUESTION 1:</b> What categories of passengers were most likely to survive the Titanic disaster? -> 4 points 
(the mark includes 2 points for the answer to the main questions and 2 points for the detailed
explanation of the logic of the analysis)

<ul>
    <li>We have seen in our data that children were more likely to survive than adults.</li>
    <li>We have seen in our data that females were far more likely to survive than males.</li>
    <li>We have seen in our data that upper and middle class passengers were far more likely to survive than lower class passengers.</li>
    <li>We have seen in our data that most passengers came from Southampton, thus so did most of the survivors.</li>
    <li>We have seen in our data that females in upper or middle class had a far greater chance of survival than any other category of passenger.</li>
</ul>

The logic of the analysis was described in great deal throughout this notebook, interspersed with the pandas code that implemented the analysis.

<b>QUESTION 2:</b> What other attributes did you use for the analysis? Explain how you used them. Provide a 
complete list of all attributes used. -> 2 points

The attributes used in my analysis included:

<ol>
	<li>PassengerID - This was used as the index for the DateFrame, which is useful for tying transformed data back to the original record in the file, if necessary.</li>
    <li>Survived - Used to determine which passengers survived the disaster and which ones did not.</li>
	<li>Pclass - Used to determine the class of cabin occupied by each passenger.</li>
	<li>Name - Used to determine the passenger's age category based on which honorific is used, when no age value was available.</li>
	<li>Sex - Used as a passenger category, as-is</li>
	<li>Age - Used to derive the passenger's age category, when a value was available</li>
	<li>Sibsp - Used to calculate the number of travelling companions accompanying the passenger</li>
	<li>Parch - Used to calculate the number of travelling companions accompanying the passenger</li>
    <li>Embarked - Used to determine if port of embarkment influenced survivorship</li>
</ol>

<b>QUESTION 3:</b> Did you engineer any attributes? If yes, explain the rationale and how the new attributes 
were used in the analysis? If you have excluded any attributes from the analysis, provide an explanation 
why you believe they have to be excluded -> 2 points

The following attributes were engineered and used in my analysis:

<ol>
	<li>SES (socio-economic status) - This attribute was engineered to provide a descriptive label for the Pclass value, for more readable output.</li>
    <li>Port - This attribute was engineered to provide a descriptive label for the Embarkement value, for more readable output.</li>
	<li>AgeCategory - This attribute was engineered to indicate whether the passenger was an adult or a child. It allowed the analysis to compensate for missing age values.</li>
	<li>Companions - This attribute was engineered to indicate the number of travelling companions for the passenger. It was used as an input to the calculation of AgeCategory.</li>
</ol>

The following attributes were excluded from the analysis:

<ol>
    <li>Cabin - This attribute was very sparsely populated, thus was omitted from the analysis.</li>
    <li>Ticket - This attribute did not provide any compelling information for this analysis.</li>
    <li>Fare - This is another socio-economic indicator, but it was made redundant by Pclass.</li>
</ol>

<b>QUESTION 4:</b> How did you treat missing values? Provide a detailed explanation in the comments. -> 2 points

In the case of the Cabin attribute, there were enough missing values that the entire attribute didn't seem useful. In this case, the attribute was dropped from the DataFrame and the analysis.

In the case of the Age attribute, an AgeCategory attribute was engineered to reflect the Age information at a summary level. The value for AgeCategory was derived from not only Age, but also from Name, Parch and Sibsp. These additional inputs allowed AgeCategory to be much more densely populated than Age, for a more complete sample to analyze.